apps/homescreen/test/unit/app_test.js

```diff
@@ -1,36 +1,564 @@
+        return Promise.resolve([
+          { id: 'abc/', data: {} },
+          { id: 'def/', data: {} }
+        ]);
+      });
+      MockNavigatormozApps.mApps = [{
+        id: 0,
+        manifest: {},
+        manifestURL: 'abc',
+        order: 0
+      }];
+
+      app = new App();
+    });
+  });
+
+  suite('App#saveSettings()', () => {
+    test('should restore data', () => {
+      app.small = true;
+      app.saveSettings();
+      app.small = false;
+      app.restoreSettings();
+
+      assert.equal(app.small, true);
+    });
+
+    test('should not restore data if version number differs', () => {
+      app.small = true;
+      app.saveSettings();
+      var tmp = JSON.parse(mockLocalStorage.mRawContent.settings);
+      tmp.version = 'UnknownVersionNumber';
+      mockLocalStorage.mRawContent.settings = JSON.stringify(tmp.settings);
+      app.small = false;
+      app.restoreSettings();
+
+      assert.equal(app.small, false);
+    });
+  });
+
+  suite('App#addApp()', () => {
+    var addAppIconStub;
+
+    setup(() => {
+      addAppIconStub = sinon.stub(app, 'addAppIcon');
+    });
+
+    teardown(() => {
+      addAppIconStub.restore();
+    });
+
+    test('should only accept valid apps', () => {
+      app.addApp({});
+      assert.isFalse(addAppIconStub.called, 'Empty app objects');
+
+      app.addApp({ manifest: { role: 'system' } });
+      assert.isFalse(addAppIconStub.called, 'Excluded role app');
+
+      app.addApp({ updateManifest: { role: 'system' } });
+      assert.isFalse(addAppIconStub.called, 'Excluded role app');
+
+      app.addApp({ manifest: {} });
+      assert.isTrue(addAppIconStub.calledOnce);
+    });
+
+    test('should add all entry points', () => {
+      app.addApp({ manifest: { entry_points: { a: {}, b: {}, c: {} } } });
+      assert.isTrue(addAppIconStub.calledThrice);
+    });
+
+    test('should be called on app install', () => {
+      var addAppStub = sinon.stub(app, 'addApp');
+      app.handleEvent(new CustomEvent('install', {
+        detail: {
+          application: { manifest: {} }
+        }
+      }));
+      assert.isTrue(addAppStub.calledOnce);
+      addAppStub.restore();
+    });
+  });
+
+  suite('App#addIconContainer()', () => {
+    test('should return a HTML element with a order property', () => {
+      app.startupMetadata = [{ order: 1 }];
+      var container = app.addIconContainer(0);
+
+      assert.isTrue(container instanceof HTMLDivElement);
+      assert.isTrue(container instanceof HTMLElement);
+      assert.isNumber(container.order);
+    });
+
+    test('should call snapScrollPositionSub() on first child', done => {
+      var appendChildSpy = sinon.spy(app.icons, 'appendChild');
+      app.startupMetadata = [{ order: 1 }];
+      app.addIconContainer(0);
+
+      setTimeout(() => {
+        assert.isTrue(appendChildSpy.called);
+        var spyCall = appendChildSpy.getCall(0);
+        assert.equal(typeof spyCall.args[1], 'function');
```

Comment:
This test check that the callback parameter of `this.icons.appendChild` is set as a function. But it is an anonymous function. `this.snapScrollPosition(0)` is not called directly. Is that sufficient?